Map id to cid instead of multihash

rklaehn opened the pull request:
Fixes #335

dignifiedquire commented:
this really problematic, the dht only works on multihashes for provider records, so we must be able to find content based on only multihashes

rklaehn commented:
> this is really problematic, the dht only works on multihashes for provider records, so we must be able to find content based on only multihashes

Doesn't bitswap work on the level of cids? Where exactly do you need to *look up* content by multihash currently?

dignifiedquire commented:
- bitswap works with both depending on the protocol version
- when we provide records they are going to be by multihash, not by cid
- when making queries to the dht we always convert to a multihash

rklaehn commented:
> * bitswap works with both depending on the protocol version

Ah, I did not know that.

> * hey are going to be by multihash, not by cid
> * when making queries to the dht we always convert to a multihash

That just requires going from cid to multihash, not vice versa, so no big deal.


So as far as I can see there are two solutions:
- change the graph table so that the key is (code, hash_id) and the value is [(code, hash_id)]
  I would prefer to avoid that since it makes graph traversal less efficient
- keep ids as 1:1 with cids but provide a way to look up blobs by hash nonetheless. Key for id could be <hash><code>, then just look for entries with a prefix of <hash>.

I guess it depends on what we expect to be the more frequent use case - lookup by cid or by hash.

dignifiedquire commented:
@rklaehn yes

rklaehn commented:
The gist of the above is that ids are still mapped to cids, but the key for looking up ids from cids is

`<hash><code>`.

So to look up ids by hash you can just do a prefix search by hash and get all codes for which we have the hash. Just pick one that has associated data. Typically, in 99.9% of all cases, there should be exactly one?

Should we go with this? The other alternative would be to have id correspond to hash, and store (id, code) in both graph key and graph content. But I would really like to avoid that since it will make all graph traversal code slower and more cumbersome, for something that is not needed 99.9% of the time.

I think this case definitely needs to be solved one way or another before v1, because it allows you to get the store in a weird state (it is confused about links) from the outside.

dignifiedquire commented on iroh-store/src/store.rs:
```diff
@@ -71,6 +73,13 @@ fn default_blob_opts() -> Options {
     opts
 }
 
+fn id_key(cid: &Cid) -> SmallVec<[u8; 64]> {
```
Can you add a doc comment describing the format please

dignifiedquire commented on iroh-store/src/store.rs:
```diff
@@ -217,6 +226,40 @@ impl Store {
     }
 
+    #[tracing::instrument(skip(self))]
+    pub async fn get_blob_by_hash(&self, hash: &Multihash) -> Result<Option<DBPinnableSlice<'_>>> {
```
can you add a test for these two new methods as well please?

dignifiedquire commented on iroh-store/src/store.rs:
```diff
@@ -291,6 +334,38 @@ impl Store {
+    fn get_ids_for_hash(
+        &self,
+        hash: &Multihash,
+    ) -> Result<impl Iterator<Item = Result<(u64, u64)>> + '_> {
```
It is very unfortunate to return a tuple of `u64`s, maybe use a struct to avoid mixing them up? Or we could introduce a new type for `Id`.

rklaehn replied:
A newtype for Id will be quite annoying since it is used all over the place. But I will make one for Code so you can not accidentally use it as an id.

dignifiedquire reviewed (commented):
I think this is a good way to go 👍

rklaehn commented:
Addressed your PR comments.

These tests are pretty much just smoke tests. I think we need more tests, but that is unrelated to this particular issue. See https://github.com/n0-computer/iroh/issues/318
